### PR TITLE
[#12048] Submit Feedback Responses Action

### PR DESCRIPTION
**Pull request by domlimm**
Part of #12048.

**Things to highlight**

**DeadlineExtensionsLogic**
- Added `getDeadlineExtension` by `userId` and `feedbackSessionId`
- Added `deleteDeadlineExtension`

**FeedbackQuestionsLogic**
- Added `getFeedbackQuestionForQuestionNumber`

**FeedbackQuestionsDb**
- Added `getFeedbackQuestionForQuestionNumber`

**FeedbackResponseCommentsLogic**
- Added `getFeedbackResponseCommentForResponse`
- Added `updateFeedbackResponseComment`

**FeedbackResponseCommentsDb**
- Added `getFeedbackResponseCommentForResponse`
- Added `updateFeedbackResponseComment`

**FeedbackResponsesLogic**
- Added `getFeedbackResponse` by `feedbackQuestionId`, `giver` and `recipient`
- Added `updateFeedbackResponseCascade`
- Added `deleteFeedbackResponseCascade`

**FeedbackResponsesDb**
- Added `getFeedbackResponse` by `feedbackQuestionId`, `giver` and `recipient`
- Added `updateFeedbackResponseCascade`

### CODE REVIEW

**Comment by samuelfangjw** [comment]
@domlimm could you help to resolve conflicts for this PR, thanks!

File: src/it/resources/data/typicalDataBundle.json [hunk 1]
```diff
@@ -659,6 +759,84 @@
...
+          "INSTRUCTORS"
+        ],
+        "showRecipientNameTo": [
+          "INSTRUCTORS"
+        ]
+      },
+      "giver": "instr1@teammates.tmt",
+      "recipient": "student1@teammates.tmt",
+      "giverSection": {
+        "id": "00000000-0000-4000-8000-000000000202"
```
Comment (samuelfangjw) [inline src/it/resources/data/typicalDataBundle.json:792]: shouldn't be required

File: src/it/resources/data/typicalDataBundle.json [hunk 2]
```diff
@@ -271,6 +326,24 @@
...
+      "id": "00000000-0000-4000-8000-000000000704",
+      "course": {
+        "id": "idOfCourse2"
+      },
+      "name": "First feedback session",
+      "creatorEmail": "instr1@course2.tmt",
+      "instructions": "Please please fill in the following questions.",
+      "startTime": "2012-04-01T22:00:00Z",
+      "endTime": "2027-04-30T22:00:00Z",
+      "sessionVisibleFromTime": "2012-03-28T22:00:00Z",
```
Comment (samuelfangjw) [inline src/it/resources/data/typicalDataBundle.json:340]: Why do we need a new session? What's makes this session unique?

File: src/main/java/teammates/sqllogic/core/FeedbackResponsesLogic.java [hunk 3]
```diff
@@ -184,4 +197,51 @@ public boolean areThereResponsesForQuestion(UUID questionId) {
...
+     * <p>Cascade updates its associated feedback response comment
+     * (e.g. associated response ID, giverSection and recipientSection).
+     *
+     * <p>If the giver/recipient field is changed, the response is updated by recreating the response
+     * as question-giver-recipient is the primary key.
+     */
+    public FeedbackResponse updateFeedbackResponseCascade(FeedbackResponse feedbackResponse)
+            throws InvalidParametersException, EntityDoesNotExistException, EntityAlreadyExistsException {
+        FeedbackResponse oldResponse = frDb.getFeedbackResponse(feedbackResponse.getId());
+        FeedbackResponse newResponse = frDb.updateFeedbackResponse(feedbackResponse);
```
Comment (samuelfangjw) [inline src/main/java/teammates/sqllogic/core/FeedbackResponsesLogic.java:213]: These are always the same.

File: src/main/java/teammates/sqllogic/core/FeedbackResponsesLogic.java [hunk 4]
```diff
@@ -184,4 +197,51 @@ public boolean areThereResponsesForQuestion(UUID questionId) {
...
+
+        boolean isResponseIdChanged = !oldResponse.getId().equals(newResponse.getId());
```
Comment (samuelfangjw) [inline src/main/java/teammates/sqllogic/core/FeedbackResponsesLogic.java:215]: Always false if new response is created. This is a randomly generated uuid, not the same as the old id.

File: src/main/java/teammates/sqllogic/core/FeedbackResponsesLogic.java [hunk 5]
```diff
@@ -184,4 +197,51 @@ public boolean areThereResponsesForQuestion(UUID questionId) {
...
+        boolean isGiverSectionChanged = !oldResponse.getGiverSection().equals(newResponse.getGiverSection());
+        boolean isRecipientSectionChanged = !oldResponse.getRecipientSection().equals(newResponse.getRecipientSection());
```
Comment (samuelfangjw) [inline src/main/java/teammates/sqllogic/core/FeedbackResponsesLogic.java:217]: Seems like whole method needs a rework.

**Review by samuelfangjw — CHANGES_REQUESTED** [review]
This is very badly done, please have another look at it. There's a reason this wasn't assigned to you, it's best left to Qirui and Cedric who are very familiar with feedback responses.

In general, we cannot blindly follow the old flow for this. The old existingResponse/updatedResponse model doesn't work well anymore.

File: src/it/resources/data/typicalDataBundle.json (hunk 2 above)
Comment (domlimm) [reply src/it/resources/data/typicalDataBundle.json:340]: Have removed this. Wanted to have a session and question with no feedback responses at all. Now using existing data 👍🏻

**Comment by jasonqiu212** [comment]
Hi @domlimm, Can I check if this PR is ready for re-review?

**Comment by samuelfangjw** [comment]
@jasonqiu212, @domlimm, unfortunately large parts of this PR needs a rework. Will redo this another time, closing this for now.

**Pull request closed without merging**